ssz/sedes/base.py

```diff
@@ -42,143 +26,58 @@ class BaseSedes(ABC, Generic[TSerializable, TDeserialized]):
     @property
     @abstractmethod
     def is_fixed_sized(self) -> bool:
-        pass
+        ...
 
     @abstractmethod
     def get_fixed_size(self) -> int:
-        pass
+        ...
 
     #
     # Serialization
     #
     @abstractmethod
     def serialize(self, value: TSerializable) -> bytes:
-        pass
+        ...
 
     #
     # Deserialization
     #
     @abstractmethod
     def deserialize(self, data: bytes) -> TDeserialized:
-        pass
+        ...
 
     #
     # Tree hashing
     #
     @abstractmethod
-    def get_hash_tree_root(self, value: TSerializable) -> bytes:
-        pass
+    def get_hash_tree_root(self, value: TSerializable) -> Hash32:
+        ...
 
     @abstractmethod
-    def chunk_count(self) -> int:
-        pass
-
-
-TSedes = BaseSedes[Any, Any]
-
-
-class BasicSedes(BaseSedes[TSerializable, TDeserialized]):
-    def __init__(self, size: int):
-        if size <= 0:
-            raise ValueError("Length must be greater than 0")
-
-        self.size = size
-
-    #
-    # Size
-    #
-    is_fixed_sized = True
-
-    def get_fixed_size(self):
-        return self.size
-
-    #
-    # Tree hashing
-    #
-    def get_hash_tree_root(self, value: TSerializable) -> bytes:
-        serialized_value = self.serialize(value)
-        return merkleize(pack((serialized_value,)))
+    def get_hash_tree_root_and_leaves(self,
+                                      value: TSerializable,
+                                      cache: CacheObj) -> Tuple[Hash32, CacheObj]:
+        ...
 
+    @abstractmethod
     def chunk_count(self) -> int:
-        return 1
-
+        ...
 
-class BaseCompositeSedes(BaseSedes[TSerializable, TDeserialized]):
-    pass
+    @abstractmethod
+    def get_key(self, value: Any) -> bytes:
+        ...
 
 
-def _compute_fixed_size_section_length(element_sedes: Iterable[TSedes]) -> int:
-    return sum(
-        sedes.get_fixed_size() if sedes.is_fixed_sized else constants.OFFSET_SIZE
-        for sedes in element_sedes
-    )
+TSedes = BaseSedes[Any, Any]
 
 
-class CompositeSedes(BaseCompositeSedes[TSerializable, TDeserialized]):
+class BaseCompositeSedes(BaseSedes[TSerializable, TDeserialized]):
     @abstractmethod
-    def _get_item_sedes_pairs(self,
-                              value: Sequence[TSerializable],
-                              ) -> Tuple[Tuple[TSerializable, TSedes], ...]:
-        pass
-
-    def _validate_serializable(self, value: Any) -> None:
-        pass
+    def get_key(self, value: Any) -> bytes:
+        ...
 
-    def serialize(self, value: TSerializable) -> bytes:
```

Comment:
Mixing code-reorganization with refactors makes for very difficult review since the diff just shows all new lines.  Was there any logic changed in this method that I should be looking at?

Reply:
Sorry! It was moved to `sedes/basic.py` for clear code dependency.
The only changed logic is the `fixed_size_section_length_cache` part:
```python
        has_fixed_size_section_length_cache = hasattr(value, '_fixed_size_section_length_cache')
        if has_fixed_size_section_length_cache:
            if value._fixed_size_section_length_cache is None:
                fixed_size_section_length = _compute_fixed_size_section_length(element_sedes)
                value._fixed_size_section_length_cache = fixed_size_section_length
            else:
                fixed_size_section_length = value._fixed_size_section_length_cache
        else:
            fixed_size_section_length = _compute_fixed_size_section_length(element_sedes)
```
For `hasattr(value, '_fixed_size_section_length_cache')`, there is a commit that I intended to apply your `@abstractmethod` suggestion: https://github.com/hwwhww/py-ssz/commit/1956608a130cba753b9349acd17ae0248c92689b. I think it's still not clear enough for `CompositeSedes` to identify the child class. I'm inclined to rework the `serialize` function inherit flow after this PR.